extension/persistence/relational-jdbc/src/main/java/org/apache/polaris/extension/persistence/relational/jdbc/DatasourceOperations.java

```diff
@@ -0,0 +1,174 @@
+
+  private final DataSource datasource;
+
+  public DatasourceOperations(DataSource datasource) {
+    this.datasource = datasource;
+  }
+
+  public void executeScript(String scriptFilePath) throws SQLException {
+    ClassLoader classLoader = DatasourceOperations.class.getClassLoader();
+    try (Connection connection = borrowConnection();
+        Statement statement = connection.createStatement()) {
+      boolean autoCommit = connection.getAutoCommit();
+      connection.setAutoCommit(true);
+      BufferedReader reader =
+          new BufferedReader(
+              new InputStreamReader(
+                  Objects.requireNonNull(classLoader.getResourceAsStream(scriptFilePath)), UTF_8));
+      StringBuilder sqlBuffer = new StringBuilder();
+      String line;
+      while ((line = reader.readLine()) != null) {
+        line = line.trim();
+        if (!line.isEmpty() && !line.startsWith("--")) { // Ignore empty lines and comments
+          sqlBuffer.append(line).append("\n");
+          if (line.endsWith(";")) { // Execute statement when semicolon is found
+            String sql = sqlBuffer.toString().trim();
+            try {
+              int rowsUpdated = statement.executeUpdate(sql);
+              LOGGER.debug("Query {} executed {} rows affected", sql, rowsUpdated);
+            } catch (SQLException e) {
+              LOGGER.error("Error executing query {}", sql, e);
+              // re:throw this as unhandled exception
+              throw new RuntimeException(e);
+            }
+            sqlBuffer.setLength(0); // Clear the buffer for the next statement
+          }
+        }
+      }
+      connection.setAutoCommit(autoCommit);
+    } catch (IOException e) {
+      LOGGER.error("Error reading the script file", e);
+      throw new RuntimeException(e);
+    } catch (SQLException e) {
+      LOGGER.error("Error executing the script file", e);
+      throw e;
+    }
+  }
+
+  public <T, R> List<R> executeSelect(
+      @Nonnull String query,
+      @Nonnull Class<T> targetClass,
+      @Nonnull Function<T, R> transformer,
+      Predicate<R> entityFilter,
+      int limit)
+      throws SQLException {
+    try (Connection connection = borrowConnection();
+        Statement statement = connection.createStatement();
+        ResultSet s = statement.executeQuery(query)) {
+      return ResultSetToObjectConverter.collect(s, targetClass, transformer, entityFilter, limit);
+    } catch (SQLException e) {
+      LOGGER.error("Error executing query {}", query, e);
+      throw e;
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  public int executeUpdate(String query) throws SQLException {
+    try (Connection connection = borrowConnection();
+        Statement statement = connection.createStatement()) {
+      boolean autoCommit = connection.getAutoCommit();
+      connection.setAutoCommit(true);
+      int result = statement.executeUpdate(query);
+      connection.setAutoCommit(autoCommit);
```

Comment:
This is not restored on exception. 

Also, I think it would be more robust to redirect all SQL execution to `runWithinTransaction` and handle all infrastructure concerns there (in one method). All other helper methods will become much simpler, I think.